include/ensmallen_bits/nsga2/nsga2_impl.hpp

```diff
@@ -0,0 +1,393 @@
+  std::vector<MatType> population;
+  population.reserve(2 * populationSize + 1);
+
+  // Pareto fronts, initialized during non-dominated sorting.
+  std::vector<std::vector<size_t> > fronts;
+  // Initialised in CrowdingDistanceAssignment.
+  std::vector<double> crowdingDistance;
+  // Initialised during non-dominated sorting.
+  std::vector<size_t> ranks;
+
+  // Controls early termination of the optimization process.
+  bool terminate = false;
+
+  // Generate the population based on a Gaussian distribution around the given
+  // starting point.
+  for (size_t i = 0; i < populationSize; i++)
+  {
+    population.push_back(arma::randu<MatType>(iterate.n_rows,
+        iterate.n_cols) + iterate);
+  }
+
+  Info << "NSGA2 initialized successfully. Optimization started." << std::endl;
+
+  // Evaluate the fitness before optimization.
+  for (size_t i = 0; i < population.size(); i++)
+    calculatedObjectives[i] = arma::Col<double>(numObjectives);
+  EvaluateObjectives(population, objectives, calculatedObjectives);
+
+  // Iterate until maximum number of generations is obtained.
+  terminate |= Callback::BeginOptimization(*this, objectives, iterate, callbacks...);
+
+  for (size_t generation = 1; generation <= maxGenerations && !terminate; generation++)
+  {
+    Info << "NSGA2: iteration " << generation << "." << std::endl;
+    terminate |= Callback::StepTaken(*this, objectives, iterate, callbacks...);
+
+    // Create new population of candidate from the present elite population.
+    // Have P_t, generate G_t using P_t.
+    BinaryTournamentSelection(population, lowerBound, upperBound);
+
+    // Evaluate the objectives for the new population.
+    calculatedObjectives.resize(population.size());
+    for (size_t i = 0; i < population.size(); i++)
+      calculatedObjectives[i] = arma::Col<double>(numObjectives);
+    EvaluateObjectives(population, objectives, calculatedObjectives);
+
+    // Perform fast non dominated sort on P_t ∪ G_t.
+    ranks.resize(population.size());
+    FastNonDominatedSort<MatType>(fronts, ranks, calculatedObjectives);
+
+    // Perform crowding distance assignment.
+    crowdingDistance.resize(population.size());
+
+    for (size_t fNum = 0; fNum < fronts.size(); fNum++)
+      CrowdingDistanceAssignment(fronts[fNum],
+                                 crowdingDistance,
+                                 lowerBound,
+                                 upperBound);
+
+    // Sort based on crowding distance.
+    std::sort(population.begin(),
+              population.end(),
+              [this, ranks, crowdingDistance, population](MatType candidateP,
+                                                          MatType candidateQ)
+              {
+                size_t idxP, idxQ;
+                for (size_t i = 0; i < population.size(); i++)
+                {
+                  if (arma::approx_equal(population[i], candidateP, "absdiff", epsilon))
+                  {
```

Comment:
Since this is a single expression, we could probably remove the extra `{}` to improve the readability?

Reply:
Yes, earlier I had separated each of the parameters on separate lines as it was exceeding the column count. I am removing the braces.